crates/synthesis/src/discrete_basis/math.rs

```diff
@@ -0,0 +1,284 @@
+
+    let mut cross = Matrix3::zeros();
+    add_cross_prod(1., &from.cross(&to), &mut cross);
+
+    let out = Matrix3::identity() + cross + cross * cross / (1. + dot);
+    if do_checks {
+        assert_so3("rotation matrix", &out);
+    }
+    out
+}
+
+/// Compute the SO(3) matrix implementing rotation about ``axis`` for an ``angle``.
+fn so3_from_angle_axis(angle: f64, axis: &Matrix3x1<f64>) -> Matrix3<f64> {
+    let mut out = Matrix3::<f64>::identity() * angle.cos();
+    add_cross_prod(angle.sin(), axis, &mut out);
+    add_outer_prod(1. - angle.cos(), axis, &mut out);
+
+    out
+}
+
+/// Decompose the SO(3) input matrix M into a balanced group commutator, that is
+///
+///     M = V W V^T W^T
+///
+/// for two SO(3) matrices V and W. See section 4.1 of https://arxiv.org/abs/quant-ph/0505030.
+pub fn group_commutator_decomposition(
+    matrix_so3: &Matrix3<f64>,
+    do_checks: bool,
+) -> (Matrix3<f64>, Matrix3<f64>) {
+    if do_checks {
+        assert_so3("GC input", matrix_so3);
+    }
+
+    let angle = solve_decomposition_angle(matrix_so3);
+
+    let e1 = Matrix3x1::new(1., 0., 0.);
+    let e2 = Matrix3x1::new(0., 1., 0.);
+
+    let vx = so3_from_angle_axis(angle, &e1);
+    let wy = so3_from_angle_axis(angle, &e2);
+
+    let tol = 1e-15;
+    let group_comm = group_commutator(&vx, &wy);
+    let group_comm_axis = rotation_axis_from_so3(&group_comm, tol);
+    let matrix_axis = rotation_axis_from_so3(matrix_so3, tol);
+
+    let sim_matrix = rotation_matrix(&group_comm_axis, &matrix_axis, do_checks);
+    let sim_matrix_t = sim_matrix.transpose();
+
+    let v = sim_matrix * vx * sim_matrix_t;
+    let w = sim_matrix * wy * sim_matrix_t;
+
+    if do_checks {
+        assert_so3("group commutator", &group_comm);
+        assert_so3("vx", &vx);
+        assert_so3("wy", &wy);
+        assert_so3("sim_matrix", &sim_matrix);
+        assert_so3("v", &v);
+        assert_so3("w", &w);
+    }
+
+    (v, w)
+}
+
+/// Compute the L2 norm of a 3-element vector.
+fn vector_norm(mat: &Matrix3x1<f64>) -> f64 {
+    let summed = mat.iter().map(|el| el.powi(2)).sum::<f64>();
+    summed.sqrt()
+}
+
+pub(crate) fn matmul_bigcomplex(
+    a: &Matrix2<Complex<f64>>,
+    b: &Matrix2<Complex<f64>>,
+) -> Matrix2<Complex<f64>> {
+    Matrix2::new(
+        a[(0, 0)] * b[(0, 0)] + a[(0, 1)] * b[(1, 0)],
+        a[(0, 0)] * b[(0, 1)] + a[(0, 1)] * b[(1, 1)],
+        a[(1, 0)] * b[(0, 0)] + a[(1, 1)] * b[(1, 0)],
+        a[(1, 0)] * b[(0, 1)] + a[(1, 1)] * b[(1, 1)],
+    )
+}
+
+/// Compute the determinant of a 3x3 matrix.
+fn determinant(matrix: &Matrix3<f64>) -> f64 {
+    matrix[(0, 0)] * matrix[(1, 1)] * matrix[(2, 2)]
+        + matrix[(0, 1)] * matrix[(1, 2)] * matrix[(2, 0)]
+        + matrix[(0, 2)] * matrix[(1, 0)] * matrix[(2, 1)]
+        - matrix[(0, 2)] * matrix[(1, 1)] * matrix[(2, 0)]
+        - matrix[(0, 1)] * matrix[(1, 0)] * matrix[(2, 2)]
+        - matrix[(0, 0)] * matrix[(1, 2)] * matrix[(2, 1)]
```

Comment:
Why not using the existing function?

```suggestion
    matrix.determinant()
```